### PR TITLE
PS-3950 (5.5): gcc-8 compilation warnings

### PR DESCRIPTION
**Pull request by inikep**
1. Add gcc-8 to Travis config file
2. Fix the following gcc-8 warnings:
- cast between incompatible function types [-Werror=cast-function-type]
- clearing an object with no trivial copy-assignment [-Werror=class-memaccess]
- directive output may be truncated [-Werror=format-truncation=]

### CODE REVIEW

File: client/mysql_plugin.c [hunk 1]
```diff
@@ -1168,7 +1170,9 @@ static int dump_bootstrap_file(char *bootstrap_file)
 
 static int bootstrap_server(char *server_path, char *bootstrap_file)
 {
-  char bootstrap_cmd[FN_REFLEN];
+  const char format_cmd[]= "%s --no-defaults --bootstrap --datadir=%s "
+                           "--basedir=%s < %s";
+  char bootstrap_cmd[FN_REFLEN * 2 + sizeof(format_cmd)];
```
Comment (percona-ysorokin) [inline client/mysql_plugin.c]: `FN_REFLEN * 4` ?

File: sql/debug_sync.cc [hunk 2]
```diff
@@ -939,8 +939,8 @@ static void debug_sync_remove_action(st_debug_sync_control *ds_control,
       produced by the shift. Again do not use an assignment operator to
       avoid string allocation/copy.
     */
-    memmove(ds_control->ds_action + ds_control->ds_active, save_action,
-            sizeof(st_debug_sync_action));
+    memmove(static_cast<void*>(ds_control->ds_action + ds_control->ds_active),
+	    save_action, sizeof(st_debug_sync_action));
```
Comment (percona-ysorokin) [inline sql/debug_sync.cc]: Tab character instead of spaces. Also need one more space for indentation.

File: client/mysql_plugin.c [hunk 3]
```diff
@@ -315,9 +315,11 @@ static char *add_quotes(const char *path)
 
 static int get_default_values()
 {
+  const char format_cmd[]= "%s mysqld > %s";
```
Comment (laurynas-biveinis) [inline client/mysql_plugin.c]: static const char

File: storage/innobase/include/data0type.ic [hunk 4]
```diff
@@ -446,6 +446,7 @@ dtype_get_fixed_size_low(
 			return(0);
 		}
 #endif /* UNIV_DEBUG */
+	  // fallthrough
```
Comment (percona-ysorokin) [inline storage/innobase/include/data0type.ic]: Indentation.

File: storage/innobase/include/data0type.ic [hunk 5]
```diff
@@ -524,6 +525,7 @@ dtype_get_min_size_low(
...
```
Comment (percona-ysorokin) [inline storage/innobase/include/data0type.ic]: Indentation.

File: client/mysql_plugin.c [hunk 6]
```diff
@@ -1168,7 +1170,9 @@ static int dump_bootstrap_file(char *bootstrap_file)
 
 static int bootstrap_server(char *server_path, char *bootstrap_file)
 {
-  char bootstrap_cmd[FN_REFLEN];
+  const char format_cmd[]= "%s --no-defaults --bootstrap --datadir=%s "
```
Comment (laurynas-biveinis) [inline client/mysql_plugin.c]: static const char

File: client/mysqlcheck.c [hunk 7]
```diff
@@ -762,7 +762,9 @@ static int handle_request_for_tables(char *tables, size_t length)
   }
   if (mysql_real_query(sock, query, query_length))
   {
-    sprintf(message, "when executing '%s TABLE ... %s'", op, options);
+    const char format_msg[]= "when executing '%s TABLE ... %s'";
```
Comment (laurynas-biveinis) [inline client/mysqlcheck.c]: static const char

File: sql/datadict.cc [hunk 8]
```diff
@@ -171,7 +171,7 @@ bool dd_recreate_table(THD *thd, const char *db, const char *table_name)
   DBUG_ASSERT(thd->mdl_context.is_lock_owner(MDL_key::TABLE, db, table_name,
                                              MDL_EXCLUSIVE));
 
-  memset(&create_info, 0, sizeof(create_info));
+  memset(static_cast<void*>(&create_info), 0, sizeof(create_info));
```
Comment (laurynas-biveinis) [inline sql/datadict.cc:174]: This does not fix the root cause of the warning. If the actual fix is too complex for us at this point of 5.5 lifecycle, just patch the build for -Wno-class-memaccess instead.

File: sql/debug_sync.cc [hunk 9]
```diff
@@ -928,7 +928,7 @@ static void debug_sync_remove_action(st_debug_sync_control *ds_control,
     memmove(save_action, action, sizeof(st_debug_sync_action));
 
     /* Move actions down. */
-    memmove(ds_control->ds_action + dsp_idx,
+    memmove(static_cast<void*>(ds_control->ds_action + dsp_idx),
```
Comment (laurynas-biveinis) [inline sql/debug_sync.cc:931]: This does not fix the root cause of the warning. If the actual fix is too complex for us at this point of 5.5 lifecycle, just patch the build for -Wno-class-memaccess instead.

File: sql/debug_sync.cc [hunk 10]
```diff
@@ -939,8 +939,8 @@ static void debug_sync_remove_action(st_debug_sync_control *ds_control,
       produced by the shift. Again do not use an assignment operator to
       avoid string allocation/copy.
     */
-    memmove(ds_control->ds_action + ds_control->ds_active, save_action,
-            sizeof(st_debug_sync_action));
+    memmove(static_cast<void*>(ds_control->ds_action + ds_control->ds_active),
```
Comment (laurynas-biveinis) [inline sql/debug_sync.cc:942]: This does not fix the root cause of the warning. If the actual fix is too complex for us at this point of 5.5 lifecycle, just patch the build for -Wno-class-memaccess instead.

File: sql/log.cc [hunk 11]
```diff
@@ -4810,7 +4810,7 @@ bool MYSQL_BIN_LOG::appendv(const char* buf, uint len,...)
 {
   bool error= 0;
   DBUG_ENTER("MYSQL_BIN_LOG::appendv");
-  va_list(args);
+  va_list args;
```
Comment (laurynas-biveinis) [inline sql/log.cc:4813]: What was the warning here?

File: sql/opt_range.cc [hunk 12]
```diff
@@ -1279,7 +1279,7 @@ QUICK_INDEX_MERGE_SELECT::QUICK_INDEX_MERGE_SELECT(THD *thd_param,
   DBUG_ENTER("QUICK_INDEX_MERGE_SELECT::QUICK_INDEX_MERGE_SELECT");
   index= MAX_KEY;
   head= table;
-  bzero(&read_record, sizeof(read_record));
+  bzero(static_cast<void*>(&read_record), sizeof(read_record));
```
Comment (laurynas-biveinis) [inline sql/opt_range.cc:1282]: This does not fix the root cause of the warning. If the actual fix is too complex for us at this point of 5.5 lifecycle, just patch the build for -Wno-class-memaccess instead.

File: sql/partition_info.cc [hunk 13]
```diff
@@ -41,7 +41,7 @@ partition_info *partition_info::get_clone(bool reset /* = false */)
     mem_alloc_error(sizeof(partition_info));
     return NULL;
   }
-  memcpy(clone, this, sizeof(partition_info));
+  memcpy(static_cast<void*>(clone), this, sizeof(partition_info));
```
Comment (laurynas-biveinis) [inline sql/partition_info.cc:44]: This does not fix the root cause of the warning. If the actual fix is too complex for us at this point of 5.5 lifecycle, just patch the build for -Wno-class-memaccess instead.

File: sql/partition_info.cc [hunk 14]
```diff
@@ -54,7 +54,7 @@ partition_info *partition_info::get_clone(bool reset /* = false */)
       mem_alloc_error(sizeof(partition_element));
       return NULL;
     }
-    memcpy(part_clone, part, sizeof(partition_element));
+    memcpy(static_cast<void*>(part_clone), part, sizeof(partition_element));
```
Comment (laurynas-biveinis) [inline sql/partition_info.cc:57]: This does not fix the root cause of the warning. If the actual fix is too complex for us at this point of 5.5 lifecycle, just patch the build for -Wno-class-memaccess instead.

File: sql/partition_info.cc [hunk 15]
```diff
@@ -84,7 +84,7 @@ partition_info *partition_info::get_clone(bool reset /* = false */)
         mem_alloc_error(sizeof(partition_element));
         return NULL;
       }
-      memcpy(subpart_clone, subpart, sizeof(partition_element));
+      memcpy(static_cast<void*>(subpart_clone), subpart, sizeof(partition_element));
```
Comment (laurynas-biveinis) [inline sql/partition_info.cc:87]: This does not fix the root cause of the warning. If the actual fix is too complex for us at this point of 5.5 lifecycle, just patch the build for -Wno-class-memaccess instead.

File: sql/partition_info.cc [hunk 16]
```diff
@@ -1361,7 +1361,7 @@ void partition_info::print_no_partition_found(TABLE *table_arg)
   char *buf_ptr= (char*)&buf;
   TABLE_LIST table_list;
 
-  bzero(&table_list, sizeof(table_list));
+  bzero(static_cast<void*>(&table_list), sizeof(table_list));
```
Comment (laurynas-biveinis) [inline sql/partition_info.cc:1364]: This does not fix the root cause of the warning. If the actual fix is too complex for us at this point of 5.5 lifecycle, just patch the build for -Wno-class-memaccess instead.

File: client/mysqlcheck.c [hunk 17]
```diff
@@ -762,7 +762,9 @@ static int handle_request_for_tables(char *tables, size_t length)
...
+    char message[sizeof(options) + sizeof(format_msg)];
```
Comment (percona-ysorokin) [inline client/mysqlcheck.c]: Strictly speaking `sizeof(options) + sizeof(format_msg) + sizeof(<max_op_len>)`, where `<op>` is one of the following:
```
CHECK
REPAIR
REPAIR NO_WRITE_TO_BINLOG
ANALYZE
ANALYZE NO_WRITE_TO_BINLOG
OPTIMIZE
OPTIMIZE NO_WRITE_TO_BINLOG
```

File: sql/sql_analyse.cc [hunk 18]
```diff
@@ -295,13 +295,12 @@ bool get_ev_num_info(EV_NUM_INFO *ev_info, NUM_INFO *info, const char *num)
   return 1;
 } // get_ev_num_info
 
-
-void free_string(String *s)
+void free_string(void* key, TREE_FREE action __attribute__((unused)),
```
Comment (laurynas-biveinis) [inline sql/sql_analyse.cc:298]: `s/__attribute__/MY_ATTRIBUTE/g`

File: sql/sql_base.cc [hunk 19]
```diff
@@ -1193,7 +1193,7 @@ bool close_cached_connection_tables(THD *thd, LEX_STRING *connection)
   DBUG_ENTER("close_cached_connections");
   DBUG_ASSERT(thd);
 
-  bzero(&tmp, sizeof(TABLE_LIST));
+  bzero(static_cast<void*>(&tmp), sizeof(TABLE_LIST));
```
Comment (laurynas-biveinis) [inline sql/sql_base.cc:1196]: This does not fix the root cause of the warning. If the actual fix is too complex for us at this point of 5.5 lifecycle, just patch the build for -Wno-class-memaccess instead.

File: sql/sql_insert.cc [hunk 20]
```diff
@@ -4078,7 +4078,7 @@ select_create::binlog_show_create_table(TABLE **tables, uint count)
   int result;
   TABLE_LIST tmp_table_list;
 
-  memset(&tmp_table_list, 0, sizeof(tmp_table_list));
+  memset(static_cast<void*>(&tmp_table_list), 0, sizeof(tmp_table_list));
```
Comment (laurynas-biveinis) [inline sql/sql_insert.cc:4081]: This does not fix the root cause of the warning. If the actual fix is too complex for us at this point of 5.5 lifecycle, just patch the build for -Wno-class-memaccess instead.

File: sql/sql_select.cc [hunk 21]
```diff
@@ -1779,7 +1779,7 @@ void JOIN::restore_tmp()
 {
   DBUG_PRINT("info", ("restore_tmp this %p tmp_join %p", this, tmp_join));
   DBUG_ASSERT(tmp_join != this);
-  memcpy(tmp_join, this, (size_t) sizeof(JOIN));
+  memcpy(static_cast<void*>(tmp_join), this, (size_t) sizeof(JOIN));
```
Comment (laurynas-biveinis) [inline sql/sql_select.cc:1782]: This does not fix the root cause of the warning. If the actual fix is too complex for us at this point of 5.5 lifecycle, just patch the build for -Wno-class-memaccess instead.

File: storage/myisam/mi_write.c [hunk 22]
```diff
@@ -943,19 +945,20 @@ static int keys_free(uchar *key, TREE_FREE mode, bulk_insert_param *param)
...
-    return 0;
+    return;
   case free_free:
     keyinfo=param->info->s->keyinfo+param->keynr;
     keylen=_mi_keylength(keyinfo, key);
     memcpy(lastkey, key, keylen);
-    return _mi_ck_write_btree(param->info,param->keynr,lastkey,
-			      keylen - param->info->s->rec_reflength);
+    _mi_ck_write_btree(param->info,param->keynr,lastkey,
+                      keylen - param->info->s->rec_reflength);
```
Comment (percona-ysorokin) [inline storage/myisam/mi_write.c]: Indentation

File: sql/sql_select.cc [hunk 23]
```diff
@@ -11106,8 +11106,8 @@ TABLE *create_virtual_tmp_table(THD *thd, List<Create_field> &field_list)
                         NullS))
     return 0;
 
-  bzero(table, sizeof(*table));
-  bzero(share, sizeof(*share));
+  bzero(static_cast<void*>(table), sizeof(*table));
```
Comment (laurynas-biveinis) [inline sql/sql_select.cc:11109]: This does not fix the root cause of the warning. If the actual fix is too complex for us at this point of 5.5 lifecycle, just patch the build for -Wno-class-memaccess instead.

File: sql/sql_select.cc [hunk 24]
```diff
@@ -11106,8 +11106,8 @@ TABLE *create_virtual_tmp_table(THD *thd, List<Create_field> &field_list)
...
+  bzero(static_cast<void*>(share), sizeof(*share));
```
Comment (laurynas-biveinis) [inline sql/sql_select.cc:11110]: This does not fix the root cause of the warning. If the actual fix is too complex for us at this point of 5.5 lifecycle, just patch the build for -Wno-class-memaccess instead.

File: sql/sql_show.cc [hunk 25]
```diff
@@ -5882,7 +5882,7 @@ static int get_schema_views_record(THD *thd, TABLE_LIST *tables,
         {
           TABLE_LIST table_list;
           uint view_access;
-          memset(&table_list, 0, sizeof(table_list));
+          memset(static_cast<void*>(&table_list), 0, sizeof(table_list));
```
Comment (laurynas-biveinis) [inline sql/sql_show.cc:5885]: This does not fix the root cause of the warning. If the actual fix is too complex for us at this point of 5.5 lifecycle, just patch the build for -Wno-class-memaccess instead.

File: sql/sql_trigger.cc [hunk 26]
```diff
@@ -1786,7 +1786,7 @@ bool Table_triggers_list::drop_all_triggers(THD *thd, char *db, char *name)
   bool result= 0;
   DBUG_ENTER("drop_all_triggers");
 
-  bzero(&table, sizeof(table));
+  bzero(static_cast<void*>(&table), sizeof(table));
```
Comment (laurynas-biveinis) [inline sql/sql_trigger.cc:1789]: This does not fix the root cause of the warning. If the actual fix is too complex for us at this point of 5.5 lifecycle, just patch the build for -Wno-class-memaccess instead.

File: sql/sql_trigger.cc [hunk 27]
```diff
@@ -2006,7 +2006,7 @@ bool Table_triggers_list::change_table_name(THD *thd, const char *db,
   LEX_STRING *err_trigname;
   DBUG_ENTER("change_table_name");
 
-  bzero(&table, sizeof(table));
+  bzero(static_cast<void*>(&table), sizeof(table));
```
Comment (laurynas-biveinis) [inline sql/sql_trigger.cc:2009]: This does not fix the root cause of the warning. If the actual fix is too complex for us at this point of 5.5 lifecycle, just patch the build for -Wno-class-memaccess instead.

File: sql/sql_truncate.cc [hunk 28]
```diff
@@ -283,7 +283,7 @@ static bool recreate_temporary_table(THD *thd, TABLE *table)
   handlerton *table_type= table->s->db_type();
   DBUG_ENTER("recreate_temporary_table");
 
-  memset(&create_info, 0, sizeof(create_info));
+  memset(static_cast<void*>(&create_info), 0, sizeof(create_info));
```
Comment (laurynas-biveinis) [inline sql/sql_truncate.cc:286]: This does not fix the root cause of the warning. If the actual fix is too complex for us at this point of 5.5 lifecycle, just patch the build for -Wno-class-memaccess instead.

File: sql/sql_view.cc [hunk 29]
```diff
@@ -215,7 +215,7 @@ fill_defined_view_parts (THD *thd, TABLE_LIST *view)
   LEX *lex= thd->lex;
   TABLE_LIST decoy;
 
-  memcpy (&decoy, view, sizeof (TABLE_LIST));
+  memcpy(static_cast<void*>(&decoy), view, sizeof (TABLE_LIST));
```
Comment (laurynas-biveinis) [inline sql/sql_view.cc:218]: This does not fix the root cause of the warning. If the actual fix is too complex for us at this point of 5.5 lifecycle, just patch the build for -Wno-class-memaccess instead.

File: sql/sql_view.cc [hunk 30]
```diff
@@ -2025,7 +2025,7 @@ mysql_rename_view(THD *thd,
       view definition parsing or use temporary 'view_def'
       object for it.
     */
-    bzero(&view_def, sizeof(view_def));
+    bzero(static_cast<void*>(&view_def), sizeof(view_def));
```
Comment (laurynas-biveinis) [inline sql/sql_view.cc:2028]: This does not fix the root cause of the warning. If the actual fix is too complex for us at this point of 5.5 lifecycle, just patch the build for -Wno-class-memaccess instead.

File: sql/tztime.cc [hunk 31]
```diff
@@ -1523,7 +1523,7 @@ my_offset_tzs_get_key(Time_zone_offset *entry,
 static void
 tz_init_table_list(TABLE_LIST *tz_tabs)
 {
-  bzero(tz_tabs, sizeof(TABLE_LIST) * MY_TZ_TABLES_COUNT);
+  bzero(static_cast<void*>(tz_tabs), sizeof(TABLE_LIST) * MY_TZ_TABLES_COUNT);
```
Comment (laurynas-biveinis) [inline sql/tztime.cc:1526]: This does not fix the root cause of the warning. If the actual fix is too complex for us at this point of 5.5 lifecycle, just patch the build for -Wno-class-memaccess instead.

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

File: storage/myisam/mi_write.c [hunk 32]
```diff
@@ -926,7 +926,7 @@ static int keys_compare(bulk_insert_param *param, uchar *key1, uchar *key2)
 }
 
 
-static int keys_free(uchar *key, TREE_FREE mode, bulk_insert_param *param)
+static void keys_free(void* vkey, TREE_FREE mode, void *vparam)
```
Comment (percona-ysorokin) [inline storage/myisam/mi_write.c:929]: Please, also report this to upstream. I still think that calling a int-returning function cast to a void-returning function can potentially lead to code corruption.

**Review by percona-ysorokin — CHANGES_REQUESTED** [review]

File: client/mysql_plugin.c (hunk 1 above)
Comment (inikep) [reply client/mysql_plugin.c]: `opt_datadir` and `opt_basedir` have no size of FN_REFLEN, they are dynamic, so `strlen(opt_datadir)` + `strlen(opt_basedir)`

File: sql/log.cc (hunk 11 above)
Comment (inikep) [reply sql/log.cc:4813]: error: unnecessary parentheses in declaration of 'args' [-Werror=parentheses]

File: sql/sql_analyse.cc (hunk 18 above)
Comment (inikep) [reply sql/sql_analyse.cc:298]: `MY_ATTRIBUTE` is not defined in 5.5

File: storage/myisam/mi_write.c (hunk 32 above)
Comment (inikep) [reply storage/myisam/mi_write.c:929]: Reported at https://bugs.mysql.com/bug.php?id=90412

File: sql/datadict.cc (hunk 8 above)
Comment (percona-ysorokin) [reply sql/datadict.cc:174]: @laurynas-biveinis 
We considered 2 alternatives:
1. Changing `memset(ptr, 0, sizeof(*ptr));` with `*ptr= T()`.
2. Removing default constructor from the classes for which there are zeroing `memset()` calls.

Neither of them seemed safe. (1) may introduce performance issues. In (2) there is high chance that we will miss initialization of certain members in non-direct constructions (like in inherited classes, placement new operators, etc.)

I didn't like the idea of suppressing this warning globally, so, I suggested to deal with only existing problems and @inikep found the easiest way to do this with `static_cast<void*>()`. With the suggested approach existing code will be warning-free and we will still get this type of warning for the new code.

Comment (laurynas-biveinis) [reply sql/datadict.cc:174]: OK, I don't like the delta against upstream increasing but preventing new instances is important. Although the probability of that happening in 5.5 in particular is vanishingly small due to being almost-EOL.

**Review by percona-ysorokin — APPROVED** [review]
LGTM

**Review by laurynas-biveinis — APPROVED** [review]

**Comment by inikep** [comment]
https://jenkins.percona.com/view/PS%205.5/job/percona-server-5.5-param/1652/

**Pull request merged**